[Modbus] Binding restarts service after config change - keeping serial port open

Trying to figure out parameters for a modbus serial device I changed modbus.cfg several times.
Changing setting sometimes seems to restart the addon.

What I noticed was that after such a restart the addon stopped working.
Checking the logs I found entries like this:

> 19:08:04.715 [ERROR] [et.wimpi.modbus.net.SerialConnection] - open port failed: Modbus Serial Master
> 19:08:04.717 [ERROR] [ing.ModbusSlaveConnectionFactoryImpl] - connect try 1/3 error: open port failed: Modbus Serial Master. Connection SerialConnection@1e19c7d[portName=/dev/ttyUSB99,port=<null>]. Endpoint ModbusSerialSlaveEndpoint@1098b8e[portName=/dev/ttyUSB99]
> 19:08:34.754 [ERROR] [et.wimpi.modbus.net.SerialConnection] - open port failed: Modbus Serial Master
> 19:08:34.756 [ERROR] [ing.ModbusSlaveConnectionFactoryImpl] - connect try 2/3 error: open port failed: Modbus Serial Master. Connection SerialConnection@1e19c7d[portName=/dev/ttyUSB99,port=<null>]. Endpoint ModbusSerialSlaveEndpoint@1098b8e[portName=/dev/ttyUSB99]
> 19:09:04.793 [ERROR] [et.wimpi.modbus.net.SerialConnection] - open port failed: Modbus Serial Master
> 19:09:04.795 [ERROR] [ing.ModbusSlaveConnectionFactoryImpl] - connect try 3/3 error: open port failed: Modbus Serial Master. Connection SerialConnection@1e19c7d[portName=/dev/ttyUSB99,port=<null>]. Endpoint ModbusSerialSlaveEndpoint@1098b8e[portName=/dev/ttyUSB99]
> 19:09:04.796 [ERROR] [ing.ModbusSlaveConnectionFactoryImpl] - re-connect reached max tries 3, throwing last error: open port failed: Modbus Serial Master. Connection SerialConnection@1e19c7d[portName=/dev/ttyUSB99,port=<null>]. Endpoint ModbusSerialSlaveEndpoint@1098b8e[portName=/dev/ttyUSB99]
> 19:09:04.798 [ERROR] [ing.ModbusSlaveConnectionFactoryImpl] - Error connecting connection SerialConnection@1e19c7d[portName=/dev/ttyUSB99,port=<null>] for endpoint ModbusSerialSlaveEndpoint@1098b8e[portName=/dev/ttyUSB99]: open port failed: Modbus Serial Master
> 19:09:04.799 [WARN ] [.binding.modbus.internal.ModbusSlave] - ModbusSlave (eehtc): Error ...

The easiest way to see if that happend was that my adapter stopped blinking (no more traffic).

Restarting openhab2 fixes the problem.
At least until I make the next change to modbus.cfg

Environment:
Openhab2 on raspberry pi 3
Serial: FTDI USB Serial converter FT232RL
Editing of modbus.cfg was made via a windows client using the openHAB-conf share.
see: https://community.openhab.org/t/modubs-configuration-for-ee800-device/21274/14

--- a/bundles/binding/org.openhab.binding.modbus.test/src/test/java/org/openhab/binding/modbus/internal/ConfigUpdatedTestCase.java
+++ b/bundles/binding/org.openhab.binding.modbus.test/src/test/java/org/openhab/binding/modbus/internal/ConfigUpdatedTestCase.java
@@ -24,6 +24,8 @@ import org.openhab.binding.modbus.ModbusBindingProvider;
 import org.openhab.core.library.types.OnOffType;
 import org.openhab.model.item.binding.BindingConfigParseException;
 import org.osgi.service.cm.ConfigurationException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 import net.wimpi.modbus.procimg.SimpleDigitalIn;
 
@@ -55,7 +57,7 @@ public class ConfigUpdatedTestCase extends TestCaseSupport {
         this.serverType = serverType;
         // Server is a bit slower to respond than normally
         // this for the testConfigUpdatedWhilePolling
-        this.artificialServerWait = 800;
+        this.artificialServerWait = 1000; // Also remember default timeout for tcp Modbus.DEFAULT_TIMEOUT
     }
 
     @Test
@@ -108,6 +110,7 @@ public class ConfigUpdatedTestCase extends TestCaseSupport {
     @Test
     public void testConfigUpdatedWhilePolling() throws UnknownHostException, org.osgi.service.cm.ConfigurationException,
             BindingConfigParseException, InterruptedException {
+        final Logger logger = LoggerFactory.getLogger(ConfigUpdatedTestCase.class);
         // run this test only for tcp server due to the customized connection string
         Assume.assumeTrue(serverType.equals(ServerType.TCP));
         MAX_WAIT_REQUESTS_MILLIS = 10000;
@@ -124,28 +127,45 @@ public class ConfigUpdatedTestCase extends TestCaseSupport {
                 ModbusBindingProvider.TYPE_DISCRETE, null, 1, 0, 2);
         putSlaveConfigParameter(cfg, serverType, SLAVE_NAME, "updateunchangeditems", "true");
         binding.updated(cfg);
+
         configureSwitchItemBinding(2, SLAVE_NAME, 0);
+
         Thread executeOnBackground = new Thread(new Runnable() {
             @Override
             public void run() {
+                logger.info("First execution started");
                 binding.execute();
+                logger.info("First execution finished");
             }
         });
 
         executeOnBackground.start();
         Thread.sleep(100);
-        // Connection should be now open (since ~100ms passed since connection).
+        // Connection should be now open (since ~100ms passed since connection)
+        // But the first query is still on the way (since server is so slow)
+
         // Simulate config update
+        // any returned connections to the old connection pool will be closed immediately on return
         binding.updated(cfg);
 
-        // Polling should work after config update
-        binding.execute();
+        // Let the previous polling round end before entering the next round
+        // We are not really interested what would happen in the "transient period"
+        // where the old connections are ongoing at the same time as the new ones
         executeOnBackground.join();
 
-        waitForRequests(2);
-        waitForConnectionsReceived(2);
-        verifyEvents();
+        // Polling should work after config update
+        logger.info("Second execution started");
+        binding.execute();
+        logger.info("Second execution finished");
 
+        // three requests, two of those due to execute() commands in this test,
+        // one due to initial automatic execute() (when updated the first time)
+        waitForRequests(3);
+        // two connections, connection is closed on second updated(), and thus connection needs to re-initated
+        waitForConnectionsReceived(2);
+        verify(eventPublisher, times(3)).postUpdate("Item1", OnOffType.ON);
+        verify(eventPublisher, times(3)).postUpdate("Item2", OnOffType.OFF);
+        verifyNoMoreInteractions(eventPublisher);
     }
 
 }

--- a/bundles/binding/org.openhab.binding.modbus.test/src/test/java/org/openhab/binding/modbus/internal/ConfigUpdatedTestCase.java
+++ b/bundles/binding/org.openhab.binding.modbus.test/src/test/java/org/openhab/binding/modbus/internal/ConfigUpdatedTestCase.java
@@ -12,8 +12,10 @@ import static org.mockito.Mockito.*;
 
 import java.net.UnknownHostException;
 import java.util.ArrayList;
+import java.util.Dictionary;
 import java.util.List;
 
+import org.junit.Assume;
 import org.junit.Test;
 import org.junit.runner.RunWith;
 import org.junit.runners.Parameterized;
@@ -43,15 +45,22 @@ public class ConfigUpdatedTestCase extends TestCaseSupport {
 
     @SuppressWarnings("serial")
     public static class ExpectedFailure extends AssertionError {
+        public ExpectedFailure(Throwable cause) {
+            initCause(cause);
+        }
     }
 
     public ConfigUpdatedTestCase(ServerType serverType) {
         super();
         this.serverType = serverType;
+        // Server is a bit slower to respond than normally
+        // this for the testConfigUpdatedWhilePolling
+        this.artificialServerWait = 800;
     }
 
     @Test
-    public void testConfigUpdated() throws UnknownHostException, ConfigurationException, BindingConfigParseException {
+    public void testConfigUpdated() throws UnknownHostException, ConfigurationException, BindingConfigParseException,
+            org.osgi.service.cm.ConfigurationException {
         // Modbus server ("modbus slave") has two digital inputs
         spi.addDigitalIn(new SimpleDigitalIn(true));
         spi.addDigitalIn(new SimpleDigitalIn(false));
@@ -72,15 +81,71 @@ public class ConfigUpdatedTestCase extends TestCaseSupport {
             waitForConnectionsReceived(1);
         }
 
+        verifyEvents();
+    }
+
+    private void verifyEvents() throws ExpectedFailure {
+
         verify(eventPublisher, never()).postCommand(null, null);
         verify(eventPublisher, never()).sendCommand(null, null);
         try {
             verify(eventPublisher).postUpdate("Item1", OnOffType.ON);
             verify(eventPublisher).postUpdate("Item2", OnOffType.OFF);
         } catch (AssertionError e) {
-            throw new ExpectedFailure();
+            throw new ExpectedFailure(e);
         }
         verifyNoMoreInteractions(eventPublisher);
+    }
+
+    /**
+     * To verify fix for https://github.com/openhab/openhab1-addons/issues/5078
+     *
+     * @throws UnknownHostException
+     * @throws org.osgi.service.cm.ConfigurationException
+     * @throws BindingConfigParseException
+     * @throws InterruptedException
+     */
+    @Test
+    public void testConfigUpdatedWhilePolling() throws UnknownHostException, org.osgi.service.cm.ConfigurationException,
+            BindingConfigParseException, InterruptedException {
+        // run this test only for tcp server due to the customized connection string
+        Assume.assumeTrue(serverType.equals(ServerType.TCP));
+        MAX_WAIT_REQUESTS_MILLIS = 10000;
+
+        spi.addDigitalIn(new SimpleDigitalIn(true));
+        spi.addDigitalIn(new SimpleDigitalIn(false));
+
+        binding = new ModbusBinding();
+
+        // Customized connection settings, keep the connection open for 2000s, no connection retries, 300ms connection
+        // timeout
+        String connection = String.format("%s:%d:30:2000000:0:1:300", localAddress().getHostAddress(), tcpModbusPort);
+        Dictionary<String, Object> cfg = addSlave(newLongPollBindingConfig(), serverType, connection, SLAVE_NAME,
+                ModbusBindingProvider.TYPE_DISCRETE, null, 1, 0, 2);
+        putSlaveConfigParameter(cfg, serverType, SLAVE_NAME, "updateunchangeditems", "true");
+        binding.updated(cfg);
+        configureSwitchItemBinding(2, SLAVE_NAME, 0);
+        Thread executeOnBackground = new Thread(new Runnable() {
+            @Override
+            public void run() {
+                binding.execute();
+            }
+        });
+
+        executeOnBackground.start();
+        Thread.sleep(100);
+        // Connection should be now open (since ~100ms passed since connection).
+        // Simulate config update
+        binding.updated(cfg);
+
+        // Polling should work after config update
+        binding.execute();
+        executeOnBackground.join();
+
+        waitForRequests(2);
+        waitForConnectionsReceived(2);
+        verifyEvents();
+
     }
 
 }

--- a/bundles/binding/org.openhab.binding.modbus.test/src/test/java/org/openhab/binding/modbus/internal/TestCaseSupport.java
+++ b/bundles/binding/org.openhab.binding.modbus.test/src/test/java/org/openhab/binding/modbus/internal/TestCaseSupport.java
@@ -98,7 +98,7 @@ public class TestCaseSupport {
     /**
      * Max time to wait for connections/requests from client
      */
-    protected static final int MAX_WAIT_REQUESTS_MILLIS = 1000;
+    protected int MAX_WAIT_REQUESTS_MILLIS = 1000;
 
     /**
      * The server runs in single thread, only one connection is accepted at a time.

--- a/bundles/binding/org.openhab.binding.modbus/src/main/java/net/wimpi/modbus/io/ModbusSerialTransaction.java
+++ b/bundles/binding/org.openhab.binding.modbus/src/main/java/net/wimpi/modbus/io/ModbusSerialTransaction.java
@@ -177,10 +177,8 @@ public class ModbusSerialTransaction implements ModbusTransaction {
             synchronized (m_IO) {
                 int tries = 0;
 
-                // toggle the id
-                m_Request.setTransactionID(c_TransactionID.increment());
-
                 do {
+                    m_Request.setTransactionID(c_TransactionID.increment());
                     try {
                         if (m_TransDelayMS > 0) {
                             try {
@@ -189,6 +187,7 @@ public class ModbusSerialTransaction implements ModbusTransaction {
                                 logger.error("InterruptedException: {}", ex.getMessage());
                             }
                         }
+
                         // write request message
                         m_IO.writeMessage(m_Request);
                         // read response message

--- a/bundles/binding/org.openhab.binding.modbus/src/main/java/net/wimpi/modbus/io/ModbusSerialTransaction.java
+++ b/bundles/binding/org.openhab.binding.modbus/src/main/java/net/wimpi/modbus/io/ModbusSerialTransaction.java
@@ -202,16 +202,19 @@ public class ModbusSerialTransaction implements ModbusTransaction {
                                 m_Request.getTransactionID(), m_SerialCon.getParameters());
                         if (tries >= m_Retries) {
                             logger.error(
-                                    "execute reached max tries {}, throwing last error: {}. Request: {}. Serial parameters: {}",
-                                    m_Retries, e.getMessage(), m_Request, m_SerialCon.getParameters());
+                                    "execute reached max tries {}, throwing last error: {}. Request: {} (unit id {} & transaction {}). Serial parameters: {}",
+                                    m_Retries, e.getMessage(), m_Request, m_Request.getUnitID(),
+                                    m_Request.getTransactionID(), m_SerialCon.getParameters());
                             throw e;
                         }
                         Thread.sleep(m_RetryDelayMillis);
                     }
                 } while (true);
                 if (tries > 0) {
-                    logger.info("execute eventually succeeded with {} re-tries. Request: {}. Serial parameters: {}",
-                            tries, m_Request, m_SerialCon.getParameters());
+                    logger.info(
+                            "execute eventually succeeded with {} re-tries. Request: {} (unit id {} & transaction id {}). Serial parameters: {}",
+                            tries, m_Request, m_Request.getUnitID(), m_Request.getTransactionID(),
+                            m_SerialCon.getParameters());
                 }
             }
 

--- a/bundles/binding/org.openhab.binding.modbus/src/main/java/net/wimpi/modbus/io/ModbusTCPTransaction.java
+++ b/bundles/binding/org.openhab.binding.modbus/src/main/java/net/wimpi/modbus/io/ModbusTCPTransaction.java
@@ -200,9 +200,11 @@ public class ModbusTCPTransaction implements ModbusTransaction {
                     // toggle and set the id
                     m_Request.setTransactionID(c_TransactionID.increment());
                     // 3. write request, and read response
+                    logger.trace("Sending request with transaction ID {}: {}", m_Request.getTransactionID(), m_Request);
                     m_IO.writeMessage(m_Request);
                     // read response message
                     m_Response = m_IO.readResponse();
+                    logger.trace("Received response with transaction ID {}", m_Response.getTransactionID());
                     break;
                 } catch (ModbusIOException ex) {
                     tries++;
@@ -212,9 +214,9 @@ public class ModbusTCPTransaction implements ModbusTransaction {
                             m_Request.getTransactionID(), m_Connection.getAddress(), m_Connection.getPort());
                     if (tries >= m_Retries) {
                         logger.error(
-                                "execute reached max tries {}, throwing last error: {}. Request: {}. Address: {}:{}",
-                                m_Retries, ex.getMessage(), m_Request, m_Connection.getAddress(),
-                                m_Connection.getPort());
+                                "execute reached max tries {}, throwing last error: {}. Request: {} (unit id {} & transaction {}). Address: {}:{}",
+                                m_Retries, ex.getMessage(), m_Request, m_Request.getUnitID(),
+                                m_Request.getTransactionID(), m_Connection.getAddress(), m_Connection.getPort());
                         throw new ModbusIOException("Executing transaction failed (tried " + m_Retries + " times)");
                     }
                     Thread.sleep(m_RetryDelayMillis);
@@ -222,8 +224,10 @@ public class ModbusTCPTransaction implements ModbusTransaction {
             } while (true);
 
             if (tries > 0) {
-                logger.info("execute eventually succeeded with {} re-tries. Request: {}. Address: {}:{}", tries,
-                        m_Request, m_Connection.getAddress(), m_Connection.getPort());
+                logger.info(
+                        "execute eventually succeeded with {} re-tries. Request: {} (unit id {} & transaction {}). Address: {}:{}",
+                        tries, m_Request, m_Request.getUnitID(), m_Request.getTransactionID(),
+                        m_Connection.getAddress(), m_Connection.getPort());
             }
 
             // 5. deal with "application level" exceptions

--- a/bundles/binding/org.openhab.binding.modbus/src/main/java/net/wimpi/modbus/io/ModbusTCPTransaction.java
+++ b/bundles/binding/org.openhab.binding.modbus/src/main/java/net/wimpi/modbus/io/ModbusTCPTransaction.java
@@ -39,7 +39,7 @@ import net.wimpi.modbus.util.Mutex;
  */
 public class ModbusTCPTransaction implements ModbusTransaction {
 
-    private static final Logger logger = LoggerFactory.getLogger(ModbusSerialTransaction.class);
+    private static final Logger logger = LoggerFactory.getLogger(ModbusTCPTransaction.class);
 
     // class attributes
     private static AtomicCounter c_TransactionID = new AtomicCounter(Modbus.DEFAULT_TRANSACTION_ID);

--- a/bundles/binding/org.openhab.binding.modbus/src/main/java/net/wimpi/modbus/net/TCPConnectionHandler.java
+++ b/bundles/binding/org.openhab.binding.modbus/src/main/java/net/wimpi/modbus/net/TCPConnectionHandler.java
@@ -64,7 +64,6 @@ public class TCPConnectionHandler implements Runnable {
             do {
                 // 1. read the request
                 ModbusRequest request = m_Transport.readRequest();
-                logger.trace("Request:{}", request.getHexMessage());
                 ModbusResponse response = null;
 
                 // test if Process image exists
@@ -73,8 +72,8 @@ public class TCPConnectionHandler implements Runnable {
                 } else {
                     response = request.createResponse();
                 }
-                logger.debug("Request:{}", request.getHexMessage());
-                logger.debug("Response:{}", response.getHexMessage());
+                logger.debug("Request (transaction id {}): {}", request.getHexMessage(), request.getTransactionID());
+                logger.debug("Response (transaction id {}): {}", response.getHexMessage(), response.getTransactionID());
 
                 m_Transport.writeMessage(response);
             } while (true);

--- a/bundles/binding/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/internal/ModbusBinding.java
+++ b/bundles/binding/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/internal/ModbusBinding.java
@@ -163,7 +163,7 @@ public class ModbusBinding extends AbstractActiveBinding<ModbusBindingProvider> 
 
     @Override
     public void deactivate() {
-        clear();
+        clearAndClose();
     }
 
     @Override
@@ -492,11 +492,11 @@ public class ModbusBinding extends AbstractActiveBinding<ModbusBindingProvider> 
     /**
      * Clear all configuration and close all connections
      */
-    private void clear() {
+    private void clearAndClose() {
         try {
             // Closes all connections by calling destroyObject method in the ObjectFactory implementation
             if (connectionPool != null) {
-                connectionPool.clear();
+                connectionPool.close();
             }
         } catch (Exception e) {
             // Should not happen
@@ -509,7 +509,7 @@ public class ModbusBinding extends AbstractActiveBinding<ModbusBindingProvider> 
     public void updated(Dictionary<String, ?> config) throws ConfigurationException {
         try {
             // remove all known items if configuration changed
-            clear();
+            clearAndClose();
             reconstructConnectionPool();
             if (config == null) {
                 logger.debug("Got null config!");

--- a/bundles/binding/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/internal/ModbusSlave.java
+++ b/bundles/binding/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/internal/ModbusSlave.java
@@ -295,6 +295,7 @@ public abstract class ModbusSlave {
         ModbusSlaveEndpoint endpoint = getEndpoint();
         ModbusSlaveConnection connection = null;
         ModbusResponse response;
+        int requestTransactionID;
         try {
             connection = getConnection(endpoint);
             if (connection == null) {
@@ -321,11 +322,12 @@ public abstract class ModbusSlave {
                 response = transaction.getResponse();
                 logger.trace("ModbusSlave ({}): response for write (FC={}) {}", name, response.getFunctionCode(),
                         response.getHexMessage());
+                requestTransactionID = transaction.getRequest().getTransactionID();
             }
-            if ((response.getTransactionID() != transaction.getTransactionID()) && !response.isHeadless()) {
+            if ((response.getTransactionID() != requestTransactionID) && !response.isHeadless()) {
                 logger.warn(
-                        "ModbusSlave ({}): Transaction id of the response does not match request {}.  Endpoint {}. Connection: {}. Ignoring response.",
-                        name, request, endpoint, connection);
+                        "ModbusSlave ({}): Transaction id of the response ({}) does not match request {}.  Endpoint {}. Connection: {}. Ignoring response.",
+                        name, response.getTransactionID(), request, requestTransactionID, endpoint, connection);
                 throw new ModbusUnexpectedTransactionIdException();
             }
         } finally {
@@ -482,6 +484,7 @@ public abstract class ModbusSlave {
         ModbusSlaveEndpoint endpoint = getEndpoint();
         ModbusSlaveConnection connection = null;
         ModbusResponse response = null;
+        int requestTransactionID;
         try {
             connection = getConnection(endpoint);
             if (connection == null) {
@@ -507,12 +510,13 @@ public abstract class ModbusSlave {
                     throw e;
                 }
 
+                requestTransactionID = transaction.getRequest().getTransactionID();
                 response = transaction.getResponse();
             }
-            if ((response.getTransactionID() != transaction.getTransactionID()) && !response.isHeadless()) {
+            if ((response.getTransactionID() != requestTransactionID) && !response.isHeadless()) {
                 logger.warn(
-                        "ModbusSlave ({}): Transaction id of the response does not match request {}.  Endpoint {}. Connection: {}. Ignoring response.",
-                        name, request, endpoint, connection);
+                        "ModbusSlave ({}): Transaction id of the response ({}) does not match request {} id {}.  Endpoint {}. Connection: {}. Ignoring response.",
+                        name, response.getTransactionID(), request, requestTransactionID, endpoint, connection);
                 throw new ModbusUnexpectedTransactionIdException();
             }
             logger.trace("ModbusSlave ({}): response for read (FC={}) {}", name, response.getFunctionCode(),

--- a/bundles/binding/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/internal/ModbusSlave.java
+++ b/bundles/binding/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/internal/ModbusSlave.java
@@ -323,12 +323,12 @@ public abstract class ModbusSlave {
                 logger.trace("ModbusSlave ({}): response for write (FC={}) {}", name, response.getFunctionCode(),
                         response.getHexMessage());
                 requestTransactionID = transaction.getRequest().getTransactionID();
-            }
-            if ((response.getTransactionID() != requestTransactionID) && !response.isHeadless()) {
-                logger.warn(
-                        "ModbusSlave ({}): Transaction id of the response ({}) does not match request {} id {}.  Endpoint {}. Connection: {}. Ignoring response.",
-                        name, response.getTransactionID(), request, requestTransactionID, endpoint, connection);
-                throw new ModbusUnexpectedTransactionIdException();
+                if ((response.getTransactionID() != requestTransactionID) && !response.isHeadless()) {
+                    logger.warn(
+                            "ModbusSlave ({}): Transaction id of the response ({}) does not match request {} id {}.  Endpoint {}. Connection: {}. Ignoring response.",
+                            name, response.getTransactionID(), request, requestTransactionID, endpoint, connection);
+                    throw new ModbusUnexpectedTransactionIdException();
+                }
             }
         } finally {
             returnConnection(endpoint, connection);
@@ -512,12 +512,13 @@ public abstract class ModbusSlave {
 
                 requestTransactionID = transaction.getRequest().getTransactionID();
                 response = transaction.getResponse();
-            }
-            if ((response.getTransactionID() != requestTransactionID) && !response.isHeadless()) {
-                logger.warn(
-                        "ModbusSlave ({}): Transaction id of the response ({}) does not match request {} id {}.  Endpoint {}. Connection: {}. Ignoring response.",
-                        name, response.getTransactionID(), request, requestTransactionID, endpoint, connection);
-                throw new ModbusUnexpectedTransactionIdException();
+
+                if ((response.getTransactionID() != requestTransactionID) && !response.isHeadless()) {
+                    logger.warn(
+                            "ModbusSlave ({}): Transaction id of the response ({}) does not match request {} id {}.  Endpoint {}. Connection: {}. Ignoring response.",
+                            name, response.getTransactionID(), request, requestTransactionID, endpoint, connection);
+                    throw new ModbusUnexpectedTransactionIdException();
+                }
             }
             logger.trace("ModbusSlave ({}): response for read (FC={}) {}", name, response.getFunctionCode(),
                     response.getHexMessage());

--- a/bundles/binding/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/internal/ModbusSlave.java
+++ b/bundles/binding/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/internal/ModbusSlave.java
@@ -306,7 +306,7 @@ public abstract class ModbusSlave {
                 transaction.setRequest(request);
                 try {
                     logger.trace(
-                            "Executing modbus request {} using transaction {} (global transaction id before increment {}) to read data",
+                            "Executing modbus request {} using transaction {} (global transaction id before increment {}) to write data",
                             transaction.getRequest(), transaction, transaction.getTransactionID());
                     transaction.execute();
                 } catch (Exception e) {

--- a/bundles/binding/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/internal/ModbusSlave.java
+++ b/bundles/binding/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/internal/ModbusSlave.java
@@ -294,28 +294,34 @@ public abstract class ModbusSlave {
             throws ModbusConnectionException, ModbusException, ModbusUnexpectedTransactionIdException {
         ModbusSlaveEndpoint endpoint = getEndpoint();
         ModbusSlaveConnection connection = null;
+        ModbusResponse response;
         try {
             connection = getConnection(endpoint);
             if (connection == null) {
                 logger.warn("ModbusSlave ({}): not connected -- aborting request {}", name, request);
                 throw new ModbusConnectionException(endpoint);
             }
-            transaction.setRequest(request);
-            try {
-                transaction.execute();
-            } catch (Exception e) {
-                // Note, one could catch ModbusIOException and ModbusSlaveException if more detailed
-                // exception handling is required. For now, all exceptions are handled the same way with writes.
-                logger.error("ModbusSlave ({}): error when executing write request ({}): {}", name, request,
-                        e.getMessage());
-                invalidate(endpoint, connection);
-                // set connection to null such that it is not returned to pool
-                connection = null;
-                throw e;
+            synchronized (transaction) {
+                transaction.setRequest(request);
+                try {
+                    logger.trace(
+                            "Executing modbus request {} using transaction {} (global transaction id before increment {}) to read data",
+                            transaction.getRequest(), transaction, transaction.getTransactionID());
+                    transaction.execute();
+                } catch (Exception e) {
+                    // Note, one could catch ModbusIOException and ModbusSlaveException if more detailed
+                    // exception handling is required. For now, all exceptions are handled the same way with writes.
+                    logger.error("ModbusSlave ({}): error when executing write request ({}): {}", name, request,
+                            e.getMessage());
+                    invalidate(endpoint, connection);
+                    // set connection to null such that it is not returned to pool
+                    connection = null;
+                    throw e;
+                }
+                response = transaction.getResponse();
+                logger.trace("ModbusSlave ({}): response for write (FC={}) {}", name, response.getFunctionCode(),
+                        response.getHexMessage());
             }
-            ModbusResponse response = transaction.getResponse();
-            logger.trace("ModbusSlave ({}): response for write (FC={}) {}", name, response.getFunctionCode(),
-                    response.getHexMessage());
             if ((response.getTransactionID() != transaction.getTransactionID()) && !response.isHeadless()) {
                 logger.warn(
                         "ModbusSlave ({}): Transaction id of the response does not match request {}.  Endpoint {}. Connection: {}. Ignoring response.",
@@ -484,21 +490,25 @@ public abstract class ModbusSlave {
                 throw new ModbusConnectionException(endpoint);
             }
             request.setUnitID(getId());
-            transaction.setRequest(request);
+            synchronized (transaction) {
+                transaction.setRequest(request);
+                try {
+                    logger.trace(
+                            "Executing modbus request {} using transaction {} (global  transaction id before increment {}) to read data",
+                            transaction.getRequest(), transaction, transaction.getTransactionID());
+                    transaction.execute();
+                } catch (ModbusException e) {
+                    logger.error(
+                            "ModbusSlave ({}): Error getting modbus data for request {}. Error: {}. Endpoint {}. Connection: {}",
+                            name, request, e.getMessage(), endpoint, connection);
+                    invalidate(endpoint, connection);
+                    // Invalidated connections should not be returned
+                    connection = null;
+                    throw e;
+                }
 
-            try {
-                transaction.execute();
-            } catch (ModbusException e) {
-                logger.error(
-                        "ModbusSlave ({}): Error getting modbus data for request {}. Error: {}. Endpoint {}. Connection: {}",
-                        name, request, e.getMessage(), endpoint, connection);
-                invalidate(endpoint, connection);
-                // Invalidated connections should not be returned
-                connection = null;
-                throw e;
+                response = transaction.getResponse();
             }
-
-            response = transaction.getResponse();
             if ((response.getTransactionID() != transaction.getTransactionID()) && !response.isHeadless()) {
                 logger.warn(
                         "ModbusSlave ({}): Transaction id of the response does not match request {}.  Endpoint {}. Connection: {}. Ignoring response.",

--- a/bundles/binding/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/internal/ModbusSlave.java
+++ b/bundles/binding/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/internal/ModbusSlave.java
@@ -326,7 +326,7 @@ public abstract class ModbusSlave {
             }
             if ((response.getTransactionID() != requestTransactionID) && !response.isHeadless()) {
                 logger.warn(
-                        "ModbusSlave ({}): Transaction id of the response ({}) does not match request {}.  Endpoint {}. Connection: {}. Ignoring response.",
+                        "ModbusSlave ({}): Transaction id of the response ({}) does not match request {} id {}.  Endpoint {}. Connection: {}. Ignoring response.",
                         name, response.getTransactionID(), request, requestTransactionID, endpoint, connection);
                 throw new ModbusUnexpectedTransactionIdException();
             }

--- a/bundles/binding/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/internal/pooling/ModbusSlaveConnectionFactoryImpl.java
+++ b/bundles/binding/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/internal/pooling/ModbusSlaveConnectionFactoryImpl.java
@@ -190,8 +190,7 @@ public class ModbusSlaveConnectionFactoryImpl
     }
 
     public void applyEndpointPoolConfigs(Map<ModbusSlaveEndpoint, EndpointPoolConfiguration> endpointPoolConfigs) {
-        this.endpointPoolConfigs = new ConcurrentHashMap<>(
-                endpointPoolConfigs);
+        this.endpointPoolConfigs = new ConcurrentHashMap<>(endpointPoolConfigs);
     }
 
     private void tryConnect(ModbusSlaveEndpoint endpoint, PooledObject<ModbusSlaveConnection> obj,

--- a/bundles/binding/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/internal/pooling/ModbusSlaveConnectionFactoryImpl.java
+++ b/bundles/binding/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/internal/pooling/ModbusSlaveConnectionFactoryImpl.java
@@ -123,6 +123,8 @@ public class ModbusSlaveConnectionFactoryImpl
 
     @Override
     public void destroyObject(ModbusSlaveEndpoint endpoint, final PooledObject<ModbusSlaveConnection> obj) {
+        logger.trace("destroyObject for connection {} and endpoint {} -> closing the connection", obj.getObject(),
+                endpoint);
         obj.getObject().resetConnection();
     }
 

